### PR TITLE
[feature] Treebeard Wiki Menu

### PR DESCRIPTION
**Pull request by caseyrollins**
Closes [#2891](https://github.com/CenterForOpenScience/osf.io/issues/2891)

### Purpose:
- Currently, the menu for navigating between wiki pages is dynamically generated through complex logic inside the ```toc.mako``` template. This PR changes the wiki navigation menu to become an instance of Treebeard.

### Changes
- ```toc.mako``` template removed in favor of creating a Treebeard instance for the menu.
- Buttons for creating and deleting a wiki now appear in the toolbar above the menu, and are still visible when the menu is collapsed.
- Project Wiki Pages expand on page load, while Component Wiki Pages must be manually expanded.
- For non-contributors, only wiki pages with content are displayed in the menu (with the exception of the 'Home' wiki page). This change relates to [#2891](https://github.com/CenterForOpenScience/osf.io/issues/2891). 

### Screenshots
- Contributor View of Wiki Menu (with all folders expanded)
screen shot 2015-07-13 at 1 11 39 pm

- Contributor View Collapsed
screen shot 2015-07-13 at 1 11 51 pm

- Non-Contributor View of Wiki Menu (with all folders expanded)
screen shot 2015-07-15 at 11 13 05 am

### CODE REVIEW

**Comment by caseyrollins** [comment]
@GaryKriebel @lbanner -- PR is ready for review!

**Comment by GaryKriebel** [comment]
@samanehsan @caneruguz

**Comment by caneruguz** [comment]
Oh wow, this looks very cool. Will be back with review.

**Comment by lbanner** [comment]
Nice @caseyrollins

File: website/addons/wiki/templates/edit.mako [hunk 1]
```diff
@@ -39,15 +51,9 @@
            </div>
         </div>
 
-        <!-- Menu without toggle in XS size only -->
-        <div class="osf-panel visible-xs">
-            <div class="osf-panel-header"> <i class="fa fa-list"> </i>  Menu </div>
-            <div class="osf-panel-body ">
-                <%include file="wiki/templates/toc.mako"/>
-            </div>
-        </div>
```
Comment (caneruguz) [inline website/addons/wiki/templates/edit.mako]: @caseyrollins It looks like you removed the narrow size specific panel doign the toggling. Have you tested this in small screens?

File: website/static/js/wikiTreebeard.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+function resolveIcon(item) {
+    if (item.children.length > 0) {
+        if (item.open) {
+            return m('i.fa.fa-folder-open', ' ');
+        }
+        return m('i.fa.fa-folder', ' ');
+    }
+}
+
```
Comment (caneruguz) [inline website/static/js/wikiTreebeard.js]: I'm not sure why you need these functions resolveToggle(), resolveIcon() separately here too. Is the regular fangorn version not working for these? I'm not seeing that something specific is happening here, is that the case?

File: website/static/js/wikiTreebeard.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        columnTitles: function () {
+            return[{
+                title: 'Name',
+                width: '100%'
+            }]
+        },
+        onload: function(tree) {
+            var tb = this;
+            tb.wikiID = window.contextVars.wiki.wikiID;
+            tb.canEdit = window.contextVars.wiki.canEdit;
```
Comment (caneruguz) [inline website/static/js/wikiTreebeard.js]: This can potentially be confusing to have canEdit right under tb. Can you put all these into a wiki object? Like this:
```js
   tb. wiki = {
       wikiID : window.contextVars.wiki.wikiID,
       canEdit : window.contextVars.wiki.canEdit;
   }
```

File: website/static/js/wikiTreebeard.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        },
+        showFilter : false,     // Gives the option to filter by showing the filter box.
+        allowMove : false,       // Turn moving on or off.
+        hoverClass : 'fangorn-hover',
+        resolveRefreshIcon : function() {
+          return m('i.fa.fa-refresh.fa-spin');
+        }
+    };
+    var grid = new Treebeard(tbOptions);
+    expandOnLoad.call(grid.tbController);
```
Comment (caneruguz) [inline website/static/js/wikiTreebeard.js]: Should this not go under the options hook "onload"?

**Comment by caneruguz** [comment]
"No wiki content" message is nice however change style to:
- Don't use red color, use class "text-muted" instead, a muted grey color. Absence of content that is optional is not a "danger" situation. 
- Make italic.

**Comment by caneruguz** [comment]
"Buttons for creating and deleting a wiki now appear next to the toggle view buttons, as opposed to appearing within the navigation menu." 

Can you put a screenshot showing that?

**Comment by caneruguz** [comment]
Can you also put a screenshot when the window length of the browser is less than 768px?

File: website/addons/wiki/templates/edit.mako (hunk 1 above)
Comment (caseyrollins) [inline website/addons/wiki/templates/edit.mako]: I removed the narrow size panel because it used the ```toc.mako``` template, which I have removed. On narrow screens, the menu looks like this: 
screen shot 2015-06-12 at 3 20 13 pm

**Comment by caseyrollins** [comment]
"Buttons for creating and deleting a wiki now appear next to the toggle view buttons, as opposed to appearing within the navigation menu." 

Screenshot:
screen shot 2015-06-12 at 3 21 45 pm

**Comment by caseyrollins** [comment]
"Can you also put a screenshot when the window length of the browser is less than 768px?"

Screenshot with browser height @ 750 px:
screen shot 2015-06-12 at 3 30 35 pm

File: website/addons/wiki/templates/edit.mako (hunk 1 above)
Comment (caseyrollins) [inline website/addons/wiki/templates/edit.mako]: Yes! The treebeard menu remains visible on small screens. And looks like this: 

screen shot 2015-06-15 at 10 44 01 am

File: website/static/js/wikiTreebeard.js (hunk 2 above)
Comment (caseyrollins) [inline website/static/js/wikiTreebeard.js]: I adjusted the JSON data being returned to the grid so that ```resolveToggle()``` can be removed. However, the customization in ```resolveIcon()``` makes it so that a folder is only displayed next to items that have children.

**Comment by caseyrollins** [comment]
@caneruguz First pass changes done :whale: Ready for a second pass!

**Comment by caneruguz** [comment]
@caseyrollins It looks like the new and delete buttons appear on the side of the panel when closed still, I think the goal was to remove those right? 

screen shot 2015-06-23 at 12 19 17 pm

File: website/static/js/wikiTreebeard.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+                            if(item.data.page.name === 'home') {
+                                return m('a', {href: item.data.page.url}, 'Home');
+                            }
+                            if(item.data.page.wiki_content === '' && !tb.wiki.canEdit) {
+                                return [
+                                    m('h', item.data.page.name),
+                                    m('span',
+                                        [m('i', {class: 'text-muted', style: 'padding-left: 10px'}, 'No wiki content')]
+                                    )
+                                ]
```
Comment (caneruguz) [inline website/static/js/wikiTreebeard.js]: jshint shows a few missing semicolons in this file, like in here

**Comment by caseyrollins** [comment]
@caneruguz Thanks for that catch. Yes, the buttons on the collapsed menu bar should be removed.

**Comment by caneruguz** [comment]
It looks fine for me from here, @sloria can do another pass.

File: website/static/js/wikiTreebeard.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
+'use strict';
```
Comment (sloria) [inline website/static/js/wikiTreebeard.js]: The name of the JS module should correspond to the constructor being exported, i.e. rename the module to `wikiMenu`.

File: website/static/js/wikiTreebeard.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+                if (parent.data.title === 'Project Wiki Pages') {
+                    tb.updateFolder(null, parent);
+                }
+            }
+        },
+        resolveRows : function (item){
+            var tb = this;
+            var columns = [];
+
+            tb.wiki = {
```
Comment (sloria) [inline website/static/js/wikiTreebeard.js]: Why is it necessary to modify the `Treebeard` object here?

File: website/static/js/wikiTreebeard.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+                id: window.contextVars.wiki.wikiID,
```
Comment (sloria) [inline website/static/js/wikiTreebeard.js]: Reusable module should not rely on global variables because globals make testing and reusability difficult. Instead, these should be passed in as arguments to the constructor.

The only place globals are permissible is in page modules (those that end in `-page.js`).

File: website/static/js/wikiTreebeard.js [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+                            }
+                            return m('a', {href: item.data.page.url}, item.data.page.name);
+                        }
+                    }
+                );
+            }
+
+            return columns;
+        },
+        showFilter : false,     // Gives the option to filter by showing the filter box.
```
Comment (sloria) [inline website/static/js/wikiTreebeard.js]: These comments could use more specificity.

File: website/static/js/wikiTreebeard.js [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        allowMove : false,       // Turn moving on or off.
+        hoverClass : 'fangorn-hover',
+        resolveRefreshIcon : function() {
+          return m('i.fa.fa-refresh.fa-spin');
```
Comment (sloria) [inline website/static/js/wikiTreebeard.js]: Fix indentation.

File: website/static/js/pages/wiki-edit-page.js [hunk 11]
```diff
@@ -85,6 +86,15 @@ if (ctx.canEditPageName) {
 
 // Apply panels
 $(document).ready(function () {
+
+    // Treebeard Wiki Menu
+    $.ajax({
```
Comment (sloria) [inline website/static/js/pages/wiki-edit-page.js]: A loading indicator should be shown while the menu is being loaded.

File: website/addons/wiki/views.py [hunk 12]
```diff
@@ -439,3 +443,73 @@ def project_wiki_validate_name(wname, **kwargs):
             message_long='A wiki page with that name already exists.'
         ))
     return {'message': wiki_name}
+
+@must_be_valid_project
+@must_be_contributor_or_public
+def project_wiki_grid_data(auth, wname, **kwargs):
+    node = kwargs['node'] or kwargs['project']
```
Comment (sloria) [inline website/addons/wiki/views.py]: This should no longer be necessary, I believe the `node` should always refer to the node being requested.

File: website/addons/wiki/tests/test_wiki.py [hunk 13]
```diff
@@ -1106,3 +1106,118 @@ def test_format_wiki_version(self):
...
+            },
+            {
+                'title': 'Component Wiki Pages',
+                'kind': 'folder',
+                'type': 'heading',
+                'children': views.format_component_wiki_pages(node=self.project, auth=self.consolidate_auth)
+            }
+        ]
+        assert_equal(data, expected)
+
```
Comment (sloria) [inline website/addons/wiki/tests/test_wiki.py]: Remove unnecessary whitespace (but leave a single newline at the end).

File: website/static/js/wikiTreebeard.js [hunk 14]
```diff
@@ -0,0 +1,101 @@
+'use strict';
+var Fangorn = require('js/fangorn');
+var $ = require('jquery');
+var bootbox = require('bootbox');
```
Comment (sloria) [inline website/static/js/wikiTreebeard.js]: Remove any unused imports.

**Comment by sloria** [comment]
1st pass finished.

File: website/static/js/pages/wiki-edit-page.js (hunk 11 above)
Comment (caseyrollins) [inline website/static/js/pages/wiki-edit-page.js]: Yep, it's right [here](https://github.com/CenterForOpenScience/osf.io/pull/2990/files#diff-d1d4c751597f63e450870e176e67a0caR36).

**Comment by caseyrollins** [comment]
@sloria Ready for second pass :octocat:

**Comment by sloria** [comment]
This moves the "New page" button away from the menu, which isn't ideal.

Old:

osf___addons_example_wiki

New:

osf___new_wiki_menu_wiki

Any reason why we can't keep the "+" button next to "Project Wiki Pages"?

**Comment by sloria** [comment]
"Home" should always be the first page in the table of contents.

Old:

osf___new_wiki_menu_wiki

New:

osf___new_wiki_menu_wiki

**Comment by GaryKriebel** [comment]
Also, in discussions with @sloria and @JeffSpies, the navigation tree as seen below:

screen shot 2015-06-30 at 3 55 25 pm

Should either have the link icon screen shot 2015-06-30 at 3 55 42 pm designating that clicking the wiki page (e.g. Component 1 - With a WIki ) will take you away from this page

**OR**

Making the Component name an unclickable "folder" with "Home" for its home wiki underneath like you do for its parent project.

Using the above as the example,

**Component 3 - With a Wiki also**
  *Home*
  *Another component wiki*

Where *Home* and   *Another component wiki* are the links.

Steve and Jeff, did I capture that correctly?
@caneruguz any thoughts on the two approaches?

**Comment by sloria** [comment]
@GaryKriebel Yes, that is correct. I think either of the two approaches is fine. For the initial iteration, let's go the simpler route and just make the components plain text (no links).

@caseyrollins Once the above is implemented, I will do another pass.

**Comment by sloria** [comment]
I'm not sure we want to have "No Wiki Content" next to every page that is blank. It adds clutter.

image

Perhaps a tooltip that says "No Wiki Content" when you hover over the name? Ping @lbanner @GaryKriebel for feedback.

**Comment by GaryKriebel** [comment]
I believe the earlier discussion was to not have it listed if it had no content (for non-contributors)

**Comment by caseyrollins** [comment]
@GaryKriebel, sounds good. And that's a simple change to make :+1:

**Comment by sloria** [comment]
@GaryKriebel How should it display for contributors? Should it say "No wiki content"?

File: website/static/js/pages/wiki-edit-page.js [hunk 15]
```diff
@@ -14,6 +14,7 @@ require('addons/wiki/static/ace-markdown-snippets.js');
 
 var $osf = require('js/osfHelpers');
 
+var wikiMenu = require('../wikiMenu');
```
Comment (sloria) [inline website/static/js/pages/wiki-edit-page.js]: `WikiMenu` is a constructor, so it should `UpperCamelCase`.

File: website/static/js/wikiMenu.js [hunk 16]
```diff
@@ -0,0 +1,114 @@
+'use strict';
+var m = require('mithril');
+var iconmap = require('js/iconmap');
+var Treebeard = require('treebeard');
+require('../css/fangorn.css');
+
+function resolveToggle(item) {
+    var toggleMinus = m('i.fa.fa-minus', ' '),
+        togglePlus = m('i.fa.fa-plus', ' ');
```
Comment (sloria) [inline website/static/js/wikiMenu.js]: > Declare one variable per var statement

https://github.com/felixge/node-style-guide#declare-one-variable-per-var-statement via https://cosdev.readthedocs.org/en/latest/style_guides/javascript.html#style

File: website/static/js/wikiMenu.js [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+                    tb.multiselected([item]);
+                }
+                columns.push({
+                    folderIcons: true,
+                    custom: function() {
+                        if(item.data.kind === 'component') {
+                            return m('h', item.data.page.name);
+                        } else if(item.data.page.wiki_content === '' && !canEdit) {
+                            return [
+                                m('h', item.data.page.name),
```
Comment (sloria) [inline website/static/js/wikiMenu.js]: Is this a typo? What level header is this supposed to be?

**Comment by GaryKriebel** [comment]
I think that is fine.

**Comment by sloria** [comment]
There's an extra row at the bottom of the grid:

wiki

**Comment by sloria** [comment]
The collapse arrow is not using the correct cursor (it should be `pointer`).

wiki2

**Comment by caseyrollins** [comment]
@sloria @GaryKriebel I don't think the 'No wiki content' message should be shown to contributors. I feel like that would just re-introduce the 'clutter' issue that Steve pointed out for non-contributors. 

The idea behind the 'No wiki content' message for a non-contributor was that it would prevent the non-contributor from visiting a page where there is no content for them to view, but displaying the page name still gives some context to the non-contributor of what wiki pages exist within the project.

**Comment by sloria** [comment]
@GaryKriebel @lbanner Is it desired to have the Component Wiki Pages collapsed by default?

osf___wiki_wiki

**Comment by lbanner** [comment]
From my perspective, I don't see that as a problem, and it helps declutter -- but uncertain as to other expectations.

**Comment by GaryKriebel** [comment]
Ideally, saving last state after either default, but I don't believe that is an option at the moment.

**Comment by sloria** [comment]
@GaryKriebel Right, we'll defer saving previous state. So should it stay collapsed by default? Also, thoughts on Casey's comment re: "No wiki content"?

**Comment by GaryKriebel** [comment]
What about for the "Home" page... if they can't see it, they can't update it.

**Comment by caseyrollins** [comment]
@GaryKriebel, what do you mean? All of the wiki pages are visible to a contributor.

**Comment by caseyrollins** [comment]
From discussion with @GaryKriebel:

For contributors, all wiki pages are displayed, with no message about their content or lack thereof.
For non-contributors, only wiki pages with content will be displayed in the menu.

**Comment by GaryKriebel** [comment]
+1

File: website/addons/wiki/views.py [hunk 18]
```diff
@@ -439,3 +443,115 @@ def project_wiki_validate_name(wname, **kwargs):
...
+            }
+            if can_edit or wiki_content is not '':
+                pages.append(page)
+    return pages
+
+
+def format_component_wiki_pages(node, auth):
+    pages = []
+    can_edit = node.has_permission(auth.user, 'write') and not node.is_registration
+    component_wiki_pages = _serialize_wiki_toc(node, auth)
```
Comment (sloria) [inline website/addons/wiki/views.py]: Why do you need to serialize into the TOC format before serializing into the treebeard format? Can you remove the intermediate serialization step?

File: website/addons/wiki/views.py [hunk 19]
```diff
@@ -439,3 +443,115 @@ def project_wiki_validate_name(wname, **kwargs):
...
+        wiki_content = wiki_page['wiki_content'].get('wiki_content')
+        component_home_wiki = {
+            'page': {
+                'url': wiki_page['url'],
+                'name': 'Home',
+                'id': wiki_page['id'],
+                'wiki_content': wiki_content
+            }
+        }
+        if can_edit or wiki_content is not '':
```
Comment (sloria) [inline website/addons/wiki/views.py]: Would be better to check truthiness of `wiki_content` rather than compare to the empty string in case the `get` call on ~521 returns `None`.

File: website/addons/wiki/views.py [hunk 20]
```diff
@@ -439,3 +443,115 @@ def project_wiki_validate_name(wname, **kwargs):
...
+
+        for component_page in wiki_page['pages_current']:
+            if component_page['name'] != 'home':
+                nested_wiki_content = component_page['wiki_content'].get('wiki_content')
+                child = {
+                    'page': {
+                        'url': component_page['url'],
+                        'name': component_page['name'],
+                        'id': component_page['wiki_id'],
+                        'wiki_content': nested_wiki_content
```
Comment (sloria) [inline website/addons/wiki/views.py]: Is the `wiki_content` actually used on the client side? These might contain a lot of text, which will unnecessarily bloat the response body.

**Comment by sloria** [comment]
Pass finished. :vhs:

File: website/addons/wiki/views.py (hunk 20 above)
Comment (caseyrollins) [inline website/addons/wiki/views.py]: It's not used so changing ```wiki_content``` to a boolean value is a much better idea.

File: website/addons/wiki/views.py (hunk 18 above)
Comment (caseyrollins) [inline website/addons/wiki/views.py]: I'm essentially using ```_serialize_wiki_toc``` as a helper function here because it gives me the component wiki pages for the node, along with the information I need from them, instead of me repeating that logic in ```format_component_wiki_pages```.

**Comment by sloria** [comment]
Doesn't have to be done before this is merged, but could you make the menu the same height as the panels? I believe @caneruguz made some flexbox CSS classes for this.

osf___wiki_wiki

**Comment by sloria** [comment]
Discussed in person: we recently made component titles on the Files pages links, so it would be good to keep that consistent in the wiki menu.

Once that is added, this is good to merge!

**Comment by caseyrollins** [comment]
@sloria Made those changes! :lollipop:
